solidity/test/bridge/VendingMachine.test.js

```diff
@@ -0,0 +1,895 @@
+const { expect } = require("chai")
+const {
+  increaseTime,
+  to1e18,
+  to1ePrecision,
+  ZERO_ADDRESS,
+  getBlockTime,
+} = require("../helpers/contract-test-helpers")
+
+describe("VendingMachine", () => {
+  let tbtcV1
+  let tbtcV2
+  let vendingMachine
+
+  let governance
+  let tokenHolder
+  let thirdParty
+
+  const unmintFee = to1ePrecision(1, 15) // 0.001
+  const initialBalance = to1e18(5) // 5 TBTC v1
+
+  beforeEach(async () => {
+    ;[
+      deployer,
+      unmintFeeUpdateInitiator,
+      vendingMachineUpgradeInitiator,
+      governance,
+      tokenHolder,
+      thirdParty,
+    ] = await ethers.getSigners()
+
+    const TestERC20 = await ethers.getContractFactory("TestERC20")
+    tbtcV1 = await TestERC20.deploy()
+    await tbtcV1.deployed()
+
+    const TBTCToken = await ethers.getContractFactory("TBTCToken")
+    tbtcV2 = await TBTCToken.deploy()
+    await tbtcV2.deployed()
+
+    await tbtcV1.mint(tokenHolder.address, initialBalance)
+
+    const VendingMachine = await ethers.getContractFactory("VendingMachine")
+    vendingMachine = await VendingMachine.deploy(
+      tbtcV1.address,
+      tbtcV2.address,
+      unmintFee
+    )
+    await vendingMachine.deployed()
+
+    await vendingMachine.connect(deployer).transferOwnership(governance.address)
+    await vendingMachine
+      .connect(deployer)
+      .transferUnmintFeeUpdateInitiatorRole(unmintFeeUpdateInitiator.address)
+    await vendingMachine
+      .connect(deployer)
+      .transferVendingMachineUpgradeInitiatorRole(
+        vendingMachineUpgradeInitiator.address
+      )
+    await tbtcV2.connect(deployer).transferOwnership(vendingMachine.address)
+
+    await tbtcV1
+      .connect(tokenHolder)
+      .approve(vendingMachine.address, initialBalance)
+  })
+
+  describe("mint", () => {
+    context("when TBTC v1 owner has not enough tokens", () => {
+      it("should revert", async () => {
+        const amount = initialBalance.add(1)
+        await tbtcV1
+          .connect(tokenHolder)
+          .approve(vendingMachine.address, amount)
+        await expect(
+          vendingMachine.connect(tokenHolder).mint(amount)
+        ).to.be.revertedWith("Transfer amount exceeds balance")
+      })
+    })
+
+    context("when TBTC v1 owner has enough tokens", () => {
+      let tx
+
+      context("when minting entire allowance", () => {
```

Comment:
What do you think about using a similar pattern for repetitive tests as in https://github.com/keep-network/hardhat-local-networks-config/blob/fa01f96a9c9d7450a976afd1be8eacc1a6b177a3/test/project.test.ts#L157-L213

For `"when minting entire allowance"` and `"when minting part of the allowance"` the only difference is the amount.

Reply:
I like the idea though in this particular case I did not want to complicate tests given that the repetition is quite minor.